x-pack/plugins/viz_editor/pie_chart_plugin/pie_chart_plugin.tsx

```diff
@@ -0,0 +1,161 @@
+import { IndexPatternPanel } from '../public/common/components/index_pattern_panel';
+import {
+  Axis,
+  getColumnIdByIndex,
+  selectColumn,
+  UnknownVisModel,
+  VisModel,
+} from '../public/common/lib';
+import { EditorPlugin, PanelComponentProps } from '../public/editor_plugin_registry';
+
+interface PieChartPrivateState {
+  sliceAxis: Axis;
+  angleAxis: Axis;
+}
+
+type PieChartVisModel = VisModel<'pieChart', PieChartPrivateState>;
+
+function dataPanel({ visModel, onChangeVisModel }: PanelComponentProps<PieChartVisModel>) {
+  return (
+    <IndexPatternPanel
+      indexPatterns={visModel.indexPatterns}
+      onChangeIndexPatterns={indexPatterns => {
+        onChangeVisModel({ ...visModel, indexPatterns });
+      }}
+    />
+  );
+}
+
+function configPanel({ visModel, onChangeVisModel }: PanelComponentProps<PieChartVisModel>) {
+  const {
+    private: {
+      pieChart: { sliceAxis, angleAxis },
+    },
+  } = visModel;
+  return (
+    <>
+      <div className="configPanel-axis">
+        <span className="configPanel-axis-title">Slice pie by</span>
+        {sliceAxis.columns.map(col => (
+          <span>{columnSummary(selectColumn(col, visModel))}</span>
+        ))}
+      </div>
+      <div className="configPanel-axis">
+        <span className="configPanel-axis-title">Size slices by</span>
+        {angleAxis.columns.map(col => (
+          <span>{columnSummary(selectColumn(col, visModel))}</span>
+        ))}
+      </div>
+    </>
+  );
+}
+
+function toExpression(viewState: PieChartVisModel) {
+  const legacyConfig = {
+    type: 'pie',
+    addTooltip: true,
+    addLegend: true,
+    legendPosition: 'right',
+    isDonut: true,
+    labels: {
+      show: false,
+      values: true,
+      last_level: true,
+      truncate: 100,
+    },
+    dimensions: {
+      metric: {
+        accessor: 1,
+        format: {
+          id: 'number',
+        },
+        params: {},
+        aggType: 'count',
+      },
+      buckets: [
+        {
+          accessor: 0,
+          format: {
+            id: 'terms',
+            params: {
+              id: 'string',
+              otherBucketLabel: 'Other',
+              missingBucketLabel: 'Missing',
+            },
+          },
+          params: {},
+          aggType: 'terms',
+        },
+      ],
+    },
+  };
+  // TODO prob. do this on an AST object and stringify afterwards
+  // TODO actually use the stuff from the viewState
+  return `sample_data | pie_chart | kibana_pie visConfig='${JSON.stringify(legacyConfig)}'`;
+}
+
+function prefillPrivateState(visModel: UnknownVisModel) {
+  if (visModel.private.pieChart) {
+    return visModel;
+  }
+
+  // TODO we maybe need a more stable way to get these
+  const xAxisRef = getColumnIdByIndex(visModel.queries, 0, 0);
+  const yAxisRef = getColumnIdByIndex(visModel.queries, 0, 1);
+
+  if (xAxisRef && yAxisRef) {
+    return setWith(
+      clone(visModel),
```

Comment:
I'd prefer that we bundle this up into our own helper `assoc` and make it work like `lodash/fp` or `ramda`, because I have a similar requirement, and there's no point to us doing this boilerplate each time. So, this would become:

`assoc('private.pieChart', { sliceAxis: /* etc */ }, visModel)`

Reply:
Yup, makes sense. There is probably a lot which can be moved into common/lib over time. I added a helper-method `setPrivateState` which also improves typing (which is bad with set, as Tim mentioned)